staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -762,14 +762,35 @@ func shouldDelegateList(opts storage.ListOptions) bool {
 	consistentReadFromStorage := resourceVersion == "" && !(consistentListFromCacheEnabled && requestWatchProgressSupported)
 	// Watch cache doesn't support continuations, so serve them from etcd.
 	hasContinuation := len(pred.Continue) > 0
-	// Serve paginated requests about revision "0" from watch cache to avoid overwhelming etcd.
-	hasLimit := pred.Limit > 0 && resourceVersion != "0"
+	// Serve paginated requests about revision "0" and "" (when consistent read from cache is supported)
+	// from watch cache to avoid overwhelming etcd.
+	consistentListFromCache := resourceVersion == "" && (consistentListFromCacheEnabled && requestWatchProgressSupported)
+	hasLimit := pred.Limit > 0 && !(resourceVersion == "0" || consistentListFromCache)
 	// Watch cache only supports ResourceVersionMatchNotOlderThan (default).
 	unsupportedMatch := match != "" && match != metav1.ResourceVersionMatchNotOlderThan
 
 	return consistentReadFromStorage || hasContinuation || hasLimit || unsupportedMatch
 }
 
+// shouldApplyListLimit determines whether the cacher should
+// apply a limit to an incoming LIST request.
+//
+// note that this function doesn't check RVM nor the Continuation token.
+// these parameters are validated by the shouldDelegateList function.
+//
+// as of today, the limit is applied only to requests that
+// request a consistent snapshot of data from the cache.
+func shouldApplyListLimit(opts storage.ListOptions, requestWatchProgressSupported bool) bool {
```

Comment:
Thinking more about this - instead of returning bool, can we instead just return the int (as limit to be used)?

Reply:
we could but i'm not sure if this would help because we need to ignore the limit for our legacy case.